design/aws-ecs/aws-ecs-service-addons.md

```diff
@@ -0,0 +1,108 @@
+- __Extensions self configure and self provision__: Service extensions provision their own resources as well as configuring the ECS service to use those resources. For example a load balancer addon creates its own load balancer and also attaches the ECS service to the load balancer. An App Mesh extension creates its own App Mesh virtual service and virtual node as well as configuring the task definition to have the right Envoy sidecar and proxy settings.
+- __Extensions are aware of each other__: There are no hidden "gotchas" when enabling a service extension. Extensions know of each other's existence. For example the application's container is aware that there is an Envoy proxy and that it must wait until the Envoy proxy has started before the app starts.
+- __Service class is agnostic to EC2 vs Fargate__: Rather than choosing up the capacity strategy in the class name customers just create a service, and add it to an environment. The service automatically modifies its settings as needed to adjust to the capacity available in the environment.
+
+
+## Implementation
+
+The new `Service` class acheives its goals via a method `Service.add()` This method is used to add extensions to the service. Each extension is responsible for customizing the service to support a specific feature. The following example shows how the end user instantiates a service with various feature's enabled:
+
+```js
+const myService = new Service(stack, 'my-service', {});
+myService.add(new Container({
+  cpu: 1024,
+  memoryMiB: 2048,
+  trafficPort: 80,
+  image: ContainerImage.fromRegistry('nathanpeck/greeter')
+}));
+myService.add(new AppMeshAddon({ mesh }));
+myService.add(new FireLensAddon());
+myService.add(new XRayAddon());
+myService.add(new CloudwatchAgentAddon());
+myService.add(new HttpLoadBalancerAddon());
+```
+
+Each addon is responsible for one aspect of the resulting service. The final generated service is the sum of all the features that were enabled. There are a few benefits to this approach:
+
+- Users no longer have to select their full architecture up front as with the current ECS patterns. Instead of chosing between `ApplicationLoadBalancedEc2Service` and `NetworkLoadBalancedEc2Service` up front, they can actually deploy a service, and then later add or remove an `HttpLoadBalancerAddon` from the service without needing to change the base class in their code.
+- Users can enable any number of different combinations of features which were not accessible before. They can build the same load balanced services they could before, but now they can also enable App Mesh, X-Ray, Cloudwatch Agent, and FireLens on the service at the same time.
```

Comment:
Are there any considerations we need to account for in terms of resource dependency? I see that there's a `bakeContainerDefinition` method that handles container dependencies, but just wondering if it might be necessary to dig into the specific add-ons/extensions proposed here to see if there's anything that might cause a hanging CFN deployment or something like that.

Reply:
I think with the existing extensions in this PR I have done a lot of testing to make sure that they work out of the box. It is possible that an unhealthy sidecar can cause the task to fail healthchecks and be considered unhealthy, causing a CFN deployment to hang. This shouldn't happen though, and if it does that would be a bug that we fix to make sure that the extension is stable.